rest-api.md

```diff
@@ -0,0 +1,767 @@
+```
+
+:::warning
+If API returns 425 code, it means database is not ready for work
+:::
+
+**GET** `/v1/modules/{module_id}/keys`
+
+`module_id` - staking router module contact address or id;
+
+Endpoint returns list of keys for module.
+
+Query:
+
+- `used` - filter for used/unused keys. Possible values: true/false;
+- `operatorIndex` - filter for keys of operator with index `operatorIndex`;
+
+Response:
+
+Response depends on `module type`
+
+```typescript
+class Response {
+  data: {
+    keys: Key[];
+    module: Module;
+  };
+  meta: {
+    elBlockSnapshot: ELBlockSnapshot;
+  };
+}
+
+interface HttpException {
+  statusCode: number;
+  message: string;
+}
+
+class TooEarlyResponse implements HttpException {
+  statusCode: number = 425;
+  message: string = 'Too early response';
+}
+
+class NotFoundException implements HttpException {
+  statusCode: number = 404;
+}
+```
+
+Example:
+
+```
+curl -X 'GET' \
+  'http://localhost:3000/v1/modules/1/keys?used=true&operatorIndex=1' \
+  -H 'accept: application/json'
+```
+
+:::warning
+If API returns 425 code, it means database is not ready for work
+:::
+
+**POST** `/v1/modules/{module_id}/keys/find`
+
+`module_id` - staking router module contact address or id;
+
+Returns all keys found in db.
+
+Request body:
+
+```typescript
+interface RequestBody {
+  // public keys in lowercase
+  pubkeys: string[];
+}
+```
+
+Response:
+
+```typescript
+class Response {
+  data: {
+    keys: Key[];
+    module: Module;
+  };
+  meta: {
+    elBlockSnapshot: ElBlockSnapshot;
+  };
+}
+
+interface HttpException {
+  statusCode: number;
+  message: string;
+}
+
+class TooEarlyResponse implements HttpException {
+  statusCode: number = 425;
+  message: string = 'Too early response';
+}
+
+class NotFoundException implements HttpException {
+  statusCode: number = 404;
+}
+```
+
+Example:
+
+```
+curl -X 'POST' \
+  'http://localhost:3000/v1/modules/1/keys/find' \
+  -H 'Content-Type: application/json' \
+  -d '{
+  "pubkeys": [
+    "pubkey"
+  ]
+}'
+```
+
+:::warning
+If API returns 425 code, it means database is not ready for work
+:::
+
+### /validators
+
+**GET** `/v1/modules/{module_id}/validators/validator-exits-to-prepare/{operator_id}`
+
+`module_id` - staking router module contact address or id;
+
+This endpoint will return N of oldest lido validators for earliest epoch when voluntary exit can be processed for specific node operator and specific `StakingRouter` module. Node operator will use this validators list for preparing pre-sign exit messages. API will find `used` keys of node operator and find for these public keys N oldest validators in `Validator` table. We consider active validators (`active_ongoing` status) or validators in `pending_initialized`, `pending_queued` statuses. Module tables state fetched from `EL` should be newer than `Validator` table state fetched from `CL`. Otherwise API will return error on request.
```

Comment:
Maybe simplify this?

```
Returns list of 'active_ongoing' validators ordered by index for specific node operator in staking module.
> :warning: If the Validator table data is outdate compared to the key table, the API returns an error.
```
Rest move to endpoint documentation inside code.

Reply:
Why validators are not fetched using block_number that used to fetch EL?

Reply:
I prefer to have more detailed description here. It might be useful for node operators.

Reply:
we cant fetch by block_number form CL, we need a slot

Reply:
Yeah, if we know block - we know slot number